From e81fd7f54ce753cec536e3c6deebfb6b6d556b19 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Tue, 24 Sep 2024 17:30:18 +0200
Subject: [PATCH 1260/1695] clk: stm32mp25: force ARM_DIVSEL for flexgen63
 config at 400MHz

The clkext2f frequency at 400MHZ, the default flexgen63 config,
is not supported without a divider by 2 as described in reference Manuel,
chapter 3.3 Cortex-A35 clocking:

The clock for the Cortex-A35 subsystem can be selected among:
a clock from the device clock generator (aka ck_cpu1_ext2f). The maximum
frequency on this clock is 400 MHz with a divider by two, enabled thanks
to the CA35SS_SSC_CHGCLKREQ SSC register.

In OpenSTLinux clock tree, we have flexgen63 = 400MHz,
so we force divider by 2 for ck_cpu1_ext2f clock, the CA35 bypass clock
with ARM_DIVSEL=0.

This patch also rename the function stm32mp2_a35_ss_on_hsi to
stm32mp2_a35_ss_on_bypass to be aligned with its behavior:
switch to bypass clock, clkext2f, provided by flexgen63 not always HSI.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git b0b3d951da51496d609826cf200f3fc1937a68c6 clk: stm32mp25: force ARM_DIVSEL for flexgen63 config at 400MHz]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I16fe9382abe5032ba491cf59263cf0f8866189ec
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/406611
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/stm32_sysconf.h |  3 +++
 core/drivers/clk/clk-stm32mp25.c            | 27 ++++++++++++++++++---
 2 files changed, 26 insertions(+), 4 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/stm32_sysconf.h b/core/arch/arm/plat-stm32mp2/stm32_sysconf.h
index b16d56d69..9cb8f6292 100644
--- a/core/arch/arm/plat-stm32mp2/stm32_sysconf.h
+++ b/core/arch/arm/plat-stm32mp2/stm32_sysconf.h
@@ -75,6 +75,9 @@ enum syscon_banks {
 #define A35SS_SSC_CHGCLKREQ_ARM_CHGCLKACK_MASK	BIT(1)
 #define A35SS_SSC_CHGCLKREQ_ARM_CHGCLKACK_SHIFT	U(1)
 
+#define A35SS_SSC_CHGCLKREQ_ARM_DIVSEL		BIT(16)
+#define A35SS_SSC_CHGCLKREQ_ARM_DIVSELACK	BIT(17)
+
 /*
  * A35SSC A35SS_SSC_PLL_FREQ1 register
  */
diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index 6babc90d5..e59f150f1 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -1655,14 +1655,33 @@ static void clk_stm32_debug_display_pdata(void)
 	clk_stm32_debug_display_others_dt_cfg(priv);
 }
 
-static void stm32mp2_a35_ss_on_hsi(void)
+static void stm32mp2_a35_ss_on_bypass(void)
 {
 	uint64_t timeout = 0;
+	uint32_t chgclkreq = stm32mp_syscfg_read(A35SS_SSC_CHGCLKREQ);
 
-	if (stm32mp_syscfg_read(A35SS_SSC_CHGCLKREQ) &
-	    A35SS_SSC_CHGCLKREQ_ARM_CHGCLKACK_MASK)
+	if (chgclkreq & A35SS_SSC_CHGCLKREQ_ARM_CHGCLKACK_MASK) {
 		/* Nothing to do, clock source is already set on bypass clock */
 		return;
+	}
+
+	/*
+	 * for clkext2f frequency at 400MHZ, the default flexgen63 config,
+	 * divider by 2 is required with ARM_DIVSEL=0
+	 */
+	if (chgclkreq & A35SS_SSC_CHGCLKREQ_ARM_DIVSEL) {
+		stm32mp_syscfg_write(A35SS_SSC_CHGCLKREQ,
+				     0U,
+				     A35SS_SSC_CHGCLKREQ_ARM_DIVSEL);
+		timeout = timeout_init_us(CLKSRC_TIMEOUT);
+		while (!timeout_elapsed(timeout))
+			if (!(stm32mp_syscfg_read(A35SS_SSC_CHGCLKREQ) &
+			      A35SS_SSC_CHGCLKREQ_ARM_DIVSELACK))
+				break;
+		if (stm32mp_syscfg_read(A35SS_SSC_CHGCLKREQ) &
+		    A35SS_SSC_CHGCLKREQ_ARM_DIVSELACK)
+			panic("Cannot set div on A35 bypass clock");
+	}
 
 	stm32mp_syscfg_write(A35SS_SSC_CHGCLKREQ,
 			     A35SS_SSC_CHGCLKREQ_ARM_CHGCLKREQ_EN,
@@ -1894,7 +1913,7 @@ static int clk_stm32_pll1_init(struct clk_stm32_priv *priv,
 	 * a configuration on the fly.
 	 */
 
-	stm32mp2_a35_ss_on_hsi();
+	stm32mp2_a35_ss_on_bypass();
 
 	ret = clk_stm32_pll_set_mux(priv, pll_conf->src);
 	if (ret != 0)
-- 
2.47.2

